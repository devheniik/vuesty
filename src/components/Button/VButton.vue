<script setup lang="ts">
import { computed } from 'vue'

const props = withDefaults(
  defineProps<{
    disabled?: boolean
    size?: 'small' | 'medium' | 'big'
    color?:
      | 'primary'
      | 'secondary'
      | 'light'
      | 'success'
      | 'warning'
      | 'danger'
      | 'upgrade'
      | 'tertiary'
      | 'info'
      | 'neutral'
    href?: string
    tag?: string
    to?: string | Record<string, any>
    target?: '_self' | '_blank' | '_parent' | '_top'
    loading?: boolean
  }>(),
  {
    disabled: false,
    size: 'medium',
    color: 'primary',
    fontWeight: 500,
  }
)

const component = computed<string>(() => {
  if (props.tag) {
    return props.tag
  }

  if (props.href) {
    return 'a'
  }
  if (props.to) {
    return 'router-link'
  }
  return 'button'
})

interface Emits {
  (e: 'click'): void
}

const emit = defineEmits<Emits>()

const onClick = () => {
  emit('click')
}

interface AttributesInterface {
  href?: string
  to?: string | Record<string, any>
  target?: '_self' | '_blank' | '_parent' | '_top'
}

const attrs = computed<AttributesInterface>(() => {
  const attributes: AttributesInterface = {}

  if (props.href) {
    attributes.href = props.href
  }

  if (props.to) {
    attributes.to = props.to
  }

  if (props.target) {
    attributes.target = props.target
  }

  return attributes
})
</script>

<template>
  <component
    v-bind="attrs"
    :is="component"
    :class="[{ 'v-button_disabled': disabled }, `v-button_${size}`, `v-button_${color}`, {'v-button_icon': $slots.icon && size !== 'small'}, {'v-button_icon_sm' : $slots.icon && size === 'small'}, 'v-button']"
    :disabled="disabled"
    @click="onClick">
    <div v-if="$slots['icon-left']" :class="[{'v-button_with-text':  $slots.default} ,'icon-container icon-left']">
      <slot name="icon-left" />
    </div>
    <div v-if="$slots.icon" class="icon-container">
      <slot name="icon" />
    </div>
    <slot v-if="loading" name="loadingSlot">
      <svg class="v-animated-spinner" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24">
        <circle class="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" stroke-width="4"></circle>
        <path
          class="opacity-75"
          fill="currentColor"
          d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"></path>
      </svg>
    </slot>

    <slot v-else />
    <div v-if="$slots['icon-right']" class="icon-container icon-right">
      <slot name="icon-right" />
    </div>
  </component>
</template>

<style lang="scss">
@import '../../assets/themes/main/components/button.scss';
</style>
