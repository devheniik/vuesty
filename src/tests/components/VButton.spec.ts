import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import VButton from '../../components/Button/VButton.vue'

describe('VButton', () => {
  it('renders the correct HTML', () => {
    const wrapper = mount(VButton, {
      propsData: {
        variant: 'primary',
        disabled: false,
        size: 'small',
        backgroundColor: 'primary',
        colorBold: 500,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emits a click event when the button is clicked', () => {
    const wrapper = mount(VButton, {
      propsData: {
        variant: 'primary',
        disabled: false,
        size: 'small',
        backgroundColor: 'primary',
        colorBold: 500,
      },
    })

    wrapper.find('button').trigger('click')
    expect(wrapper.emitted().click).toBeTruthy()
  })

  it('has the correct computed properties', () => {
    const wrapper = mount(VButton, {
      propsData: {
        variant: 'primary',
        disabled: false,
        size: 'small',
        backgroundColor: 'primary',
        colorBold: 500,
      },
    })

    expect(wrapper.vm.component).toBe('button')
  })
})
