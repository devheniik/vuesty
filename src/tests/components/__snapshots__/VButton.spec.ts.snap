// Vitest Snapshot v1

exports[`VButton > renders the correct HTML 1`] = `"<button class=\\"btn flex btn-primary btn-small btn-primary-500\\"></button>"`;
